packages/dds/sequence-sequential-id/src/sharedStringWithSequentialId.ts

```diff
@@ -0,0 +1,88 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Lazy } from "@fluidframework/common-utils";
+import { IChannelAttributes, IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { Marker, MergeTreeDeltaType, MergeTreeMaintenanceType,
+     reservedMarkerIdKey, SortedSegmentSet } from "@fluidframework/merge-tree";
+import { SharedString, SequenceDeltaEvent, SequenceMaintenanceEvent } from "@fluidframework/sequence";
+import { createIdBetween, digitLast, digitZero } from "./generateSequentialId";
+import { SharedStringWithSequentialIdFactory } from "./SharedStringWithSequentialIdFactory";
+
+export function sharedStringWithSequentialIdMixin(Base: typeof SharedString = SharedString): typeof SharedString {
+    return class SharedStringWithSequentialId extends Base {
+        private readonly sortedMarkers: Lazy<SortedSegmentSet>;
+        constructor(document: IFluidDataStoreRuntime, public id: string, attributes: IChannelAttributes) {
+            super(document, id, attributes);
+            this.sortedMarkers = new Lazy<SortedSegmentSet>(() => {
+                const set = new SortedSegmentSet();
+                this.walkSegments((segment) => {
+                    if (Marker.is(segment) && segment.getId() !== undefined) {
+                        // If it has an id then we have acked it
+                        set.addOrUpdate(segment);
+                    }
+                    return true;
+                });
+                return set;
+            });
+            this.on("maintenance", this.applyIdToLocalAckedSegment);
+            this.on("sequenceDelta", this.applyIdToRemoteAckedSegment);
+        }
+
+        public static getFactory() {
+            return new SharedStringWithSequentialIdFactory();
```

Comment:
That's a concreate class, while this function is essentially a generic. I.e. this is correct only if Base = SharedString. If it's not and some other mixin, then callign getFactory will return me a factory that does not know anything about this mixin.
If we want to keep this code generic, then we need to require a caller to give us a pointer to a factory (in addition to giving us Base), i.e. sharedStringWithSequentialIdMixin() needs to have two arguments. Not sure how it will work (given it a try), as it sort of creates circular reference between factory and class (though we already have circle for any factory, I guess)

Reply:
It is kind of weird... with the circular reference but added passing in the factory as a parameter. I could actually see this instead of base being sharedstring being SharedSegmentSequente or something more generic but leaving as is for now